doc.go

```diff
@@ -0,0 +1,55 @@
+package doc
+
+import (
+	"io"
+	"os/exec"
+)
+
+// JobsList stores a map of the Job's containing the linux process key'd by job id (in memory DB)
+type JobsList map[string]*Job
+
+// Job maintains the exec.Cmd struct (containing the underlying os.Process) and the "owner" for authz
+type Job struct {
+	owner string
+	cmd   *exec.Cmd
+}
+
+// JobOpts wraps the options that can be passed to cgroups for the job
+// details at https://facebookmicrosites.github.io/cgroup2/docs/overview
+type JobOpts struct {
+	CpuWeight int    // `cpu.weight`
+	MemLimit  string // `mem.high`
+	IOLatency string // `io.latency`
+}
+
+// JobStatus is an amalgamation of the useful status information available from the exec.Cmd struct of the job and it's underlying os.Process
+type JobStatus struct {
+	ID       string
+	Owner    string
+	PID      int
+	Running  bool
+	ExitCode int
+}
+
+// Worker defines the libraries API on how to start / stop / query status and get output of a job
+type Worker interface {
+	Start(opts JobOpts, owner, name string, args ...string) (id string, err error)
+	Stop(id string) error
+	Status(id string) JobStatus
+	Output(id string) (io.Reader, error)
+}
+
+// ResourceController defines the interface for implementing resource control of new processes
+// In cgroups this will be creating, editing and deleting files in /sys/fs/cgroup
+type ResourceController interface {
+	AddProcess(int) error
+	CreateGroup(string) error
+	DeleteGroup(string) error
+	AddResourceControl(JobOpts) error
+}
+
+// JobWorker implements Worker and will be initiated once by the binary starting the gRPC server
+type JobWorker struct {
```

Comment:
Does the library need this centralized struct that stores jobs? Would this functionality make more sense in the API server?

Reply:
It doesn't need to. At the very least though it would need a map of job UUID -> PID so that the library could find the process in subsequent calls such as status / stop / stream output. I did this as it was convenient for the job worker to be able to manage the job with access to the underlying exec.Cmd. I could change the API so that create returns a `Job`? This would contain the cmd.Exec and allow something like the API server to store these. One advantage of the library storing the exec.Cmd is that the caller could accidentally mutate the exec.Cmd, such as it's STDOUT and stop other functions in the library working. This way storing the list of running jobs is a way to safe guard a user who does not have implementation knowledge of the library and what it can / cannot do to exec.Cmd. But returning the Job after creation, and allowing the API server to handle this, does gives a lot of flexibility to the client.

Reply:
Looking at some other libraries online such as https://github.com/lxc/go-lxc, it seems to support the idea of returning a Job (container for lxc) after a call to Start, then allow the consumer to call Status, Stop and Output on the job itself

Reply:
I've updated doc.go now to do this and remove the in memory list of jobs as this can be implemented in the gRPC server. I've updated the example in the RFD to reflect this